### PR TITLE
`TI.log_url` fix for `map_index`

### LINKED ISSUE #24328: `TI.log_url` is incorrect with mapped tasks
### Apache Airflow version

2.3.0

### What happened

I had an `on_failure_callback` that sent a `task_instance.log_url` to slack, it no longer behaves correctly - giving me a page with no logs rendered instead of the logs for my task.

(Example of failure, URL like: https://XYZ.astronomer.run/dhp2pmdd/log?execution_date=2022-06-05T00%3A00%3A00%2B00%3A00&task_id=create_XXX_zip_files_and_upload&dag_id=my_dag )
image

### What you think should happen instead

The correct behavior would be the URL: 
https://XYZ.astronomer.run/dhp2pmdd/log?execution_date=2022-06-05T00%3A00%3A00%2B00%3A00&task_id=create_XXX_zip_files_and_upload&dag_id=my_dag&map_index=0 

as exemplified:
image

### How to reproduce

_No response_

### Operating System

Debian/Docker

### Versions of Apache Airflow Providers

_No response_

### Deployment

Astronomer

### Deployment details

_No response_

### Anything else

_No response_

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by fritz-astronomer**
closes: #24328 
related: #24328

Adds `map_index` to the `task_instance.log_url` link, which fixes a bug previously that prevented logs being displayed correctly at the link given.

### CODE REVIEW

**Comment by fritz-astronomer** [comment]
Note - my local breeze tests may not have been working correctly. I had failures but mostly with `tests/providers`

**Comment by fritz-astronomer** [comment]
This could definitely use some more unhappy-path tests. That required a slight overhaul to some of the pytest fixtures, so I didn't undertake it yet

File: tests/models/test_taskinstance.py [hunk 1]
```diff
@@ -1307,26 +1307,15 @@ def test_get_num_running_task_instances(self, create_task_instance):
...
-    #     d = urllib.parse.parse_qs(
-    #         urllib.parse.urlparse(ti.log_url).query,
-    #         keep_blank_values=True, strict_parsing=True)
-    #     self.assertEqual(d['dag_id'][0], 'dag')
-    #     self.assertEqual(d['task_id'][0], 'op')
-    #     self.assertEqual(pendulum.parse(d['execution_date'][0]), now)
-
     def test_log_url(self, create_task_instance):
-        ti = create_task_instance(dag_id='dag', task_id='op', execution_date=timezone.datetime(2018, 1, 1))
+        ti = create_task_instance(dag_id='dag', task_id='op', execution_date=timezone.datetime(2018, 1, 1), )
```
Comment (bbovenzi) [inline tests/models/test_taskinstance.py]: Why do we need the last comma for this test?

Comment (fritz-astronomer) [reply tests/models/test_taskinstance.py]: I think my brain might've just turned off 😁 I believe I was considering expanding this to include mapped tasks in the constructor, but that looked harder than I had time for. I can revert this line

File: airflow/models/taskinstance.py [hunk 2]
```diff
@@ -264,13 +263,13 @@ def clear_task_instances(
 
         drs = (
             session.query(DagRun)
-            .filter(
+                .filter(
```
Comment (bbovenzi) [inline airflow/models/taskinstance.py]: It looks like the static checks weren't happy with these indentations?

Comment (fritz-astronomer) [reply airflow/models/taskinstance.py]: Unintentional change - reverted

File: tests/models/test_taskinstance.py (hunk 1 above)
Comment (bbovenzi) [reply tests/models/test_taskinstance.py]: One more static check failure here https://github.com/apache/airflow/runs/6819118390?check_suite_focus=true#step:10:287

File: tests/models/test_taskinstance.py [hunk 3]
```diff
@@ -1307,26 +1307,19 @@ def test_get_num_running_task_instances(self, create_task_instance):
...
-    #     d = urllib.parse.parse_qs(
-    #         urllib.parse.urlparse(ti.log_url).query,
-    #         keep_blank_values=True, strict_parsing=True)
-    #     self.assertEqual(d['dag_id'][0], 'dag')
-    #     self.assertEqual(d['task_id'][0], 'op')
-    #     self.assertEqual(pendulum.parse(d['execution_date'][0]), now)
-
     def test_log_url(self, create_task_instance):
-        ti = create_task_instance(dag_id='dag', task_id='op', execution_date=timezone.datetime(2018, 1, 1))
+        ti = create_task_instance(
```
Comment (bbovenzi) [inline tests/models/test_taskinstance.py]: Another static check failure cause by moving this to multiple lines:
https://github.com/apache/airflow/runs/6819118390?check_suite_focus=true#step:10:287

**Review by uranusjr — APPROVED** [review]

**Review by bbovenzi — APPROVED** [review]

**Review by potiuk — APPROVED** [review]

**Review by jedcunningham — APPROVED** [review]

**Comment by fritz-astronomer** [comment]
build failure appears to be unrelated to my code, `release_notes.md` misspelling - I cannot merge, so if anyone else wants to click the button (or let me know if I need to find time to make those unhappy path tests 😅 )

**Review by potiuk — APPROVED** [review]

**Comment by potiuk** [comment]
Merged @fritz-astronomer :) - it was indeed intermittent failure.

**Pull request merged**